### DISCUSSION TITLE
[Proposal] Single line using namespace

### DISCUSSION THREAD

**Discussion by Gnitry** [dotnet/csharplang#3590] [Language Ideas]
Hi.
What about add ability to write namespace using in single line separated by comma?

Now:
```
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
```

Proposed:
```
using System, System.Collections.Generic, System.IO, System.Linq, System.Windows, System.Windows.Controls;
```

The reason to for this is to decrease the header size of the file. Sometimes file begins with a lot of lines of "using ...".

**Comment by SingleAccretion** [comment]
A similar proposal: #1696.

**Comment by Gnitry** [comment]
Looks similar, but nobody proposed to use just a single-line comma-separated list

**Comment by pinkfloydx33** [comment]
It's very easy to visually parse a vertical list, even one that doesn't fit on the screen. A horizontal one--especially comma separated--not so much. It would also very likely require horizontal scrolling to see the entire list especially if there's more than just a couple which by your own admission is the reason you want the feature. If there's one thing I absolutely hate when I'm *reading* code it's anything that requires me to scroll to the right (and inevitably _back_ to the left). 

While it's not difficult to split a list on commas, consider that tooling needs to constantly do this while you type. The compiler is already recognizing the semicolon to indicate the end of the statement--it would now need a new mode that has to constantly search for and then split on commas. It may not seem like much but it could potentially cause performance concerns on the tooling side. 

If you don't like the amount of space it takes up, you're free to wrap it in a `#region`

**Comment by Gnitry** [comment]
Yes, I agree with you about horizontal scrolling.
But at the same time the `#region` doesn't look like the solution for me, because it will add additional 2 lines of code, which I will see when using editor that doesn't support regions folding.

**Comment by pinkfloydx33** [comment]
Sounds like you need a better editor then. I'm not aware of a modern editor that doesn't support collapsing either natively or via a plugin/extension. That include notepad++ and VS Code. I'm not a personal fan of regions but it's a solution to the problem that doesn't involve updating the language. Unless you're using regular notepad or a terminal based editor it should be readily available and is one reason I assume this proposal is likely to be rejected.

**Comment by Gnitry** [comment]
Regarding using in terminal editors - yes, I use them. Just since started using .Net Core on embedded Linux. And I am aware about vertical scrolling the same as about horizontal scrolling.

And when I use Python I am happy that I can sometimes write
```
import os, sys, math
import superlongnamespace
```
instead of 
```
import os
import sys
import math
import superlongnamespace
```

even if it goes against Python Style Guide.
PS: for the sake of justice: Python is more ready for single-line imports, because usually they have short names

**Comment by mburbea** [comment]
I mean if you're goal is just to put it on one line you can just do this?
```cs
using System; using System.Collections.Generic; using System.IO; using System.Linq; using System.Windows;
```
Just pretend that `; using` is a comma...

**Comment by Gnitry** [comment]
```
using System; using System.Collections.Generic; using System.IO; using System.Linq; using System.Windows;
```
is not the same as
```
using System, System.Collections.Generic, System.IO, System.Linq, System.Windows;
```
So no, thanks.
My goal is to make code shorter somehow, and this is the yet another solution to do it. Far from ideal but should be discussed publicly.

**Comment by mburbea** [comment]
I don't think that's the goal of the language. You originally wanted a one-liner solution, now you want a more terse format.  I don't think theres much value add in saving 6 characters to add some new syntax that people are rarely going to use. I rarely manually modify my usings. I let visual studio/code add them as I need them and then I use cleanup tools to sort and remove the unnecessary using statements. 

Yes, there have been cases where the language team has added new syntax to save characters, but they need to be in an area where the verbosity was painful and not having it seemed unwieldy. But rarely was it for a saving of 6 characters in an area of code that most people gloss right over.

**Comment by CyrusNajmabadi** [comment]
We do care about brevity, but for the 'meat' of code.  i.e the stuff you're editing non-stop 98% of your day.  So, things like statements/expressions, and the stuff within class/method bodies. I actually don't know the last time i ever manually touched a `usings` list. It's a corner that i don't think really benefits from attention here.

**Comment by rootflood** [comment]
> @CyrusNajmabadi  I actually don't know the last time i ever manually touched a usings list.

is that real? how? because if you just create a new project then the "vs" automatically add a list of using for your main file, and the GitHub is full of "using", but however.
-this feature is important to me but i don't fight for that, the reason is clear.
-i mostly using the "using" keyword, i just imaging what if i have 10 "cs" files and every one of that have just 5 numbers of "using" keyword, then i have 50 of "using"s
-"using static" make the situation worse even.
@Gnitry 
that feature even exist in vb.net
```vb
'vb.net
Imports System, System.IO, System.Collections
``` 
but I don't know why they ain't to do that, I think  they don't like the vb

**Comment by CyrusNajmabadi** [comment]
> is that real? how? because if you just create a new project then the "vs" automatically add a list of using for your main file, and the GitHub is full of "using", but however.

Right.  But i never touch that list myself.  I do things like "add using" when i reference a new type.  Or i say "cleanup code" and VS removes the usings i'm not using anymore.  I don't need to ever edit that code myself.

--

Finally, please keep your conduct professional here.

**Comment by rootflood** [comment]
> @CyrusNajmabadi >> Finally, please keep your conduct professional here.

I did.

>Right. But i never touch that list myself.

this feature is important to me but i don't fight for that, the reason is you at last telling its not valueable.
I think this discussion isn't valueable so I don't answer any more.

**Comment by CyrusNajmabadi** [comment]
> I did.

Posting pictures like that is not professional and does nothing to aid discourse here.  Please don't do that.

**Comment by CyrusNajmabadi** [comment]
> the reason is you at last telling its not valueable.

I think you need to recognize how csharplang works.  When people create proposals, it will be discussed. Views will be shared, and community, team and LDM members will provide their own inputs on how to assess the proposals.  Those inputs may absolutely not align with yours.  That doesn't make the input non-valuable.  It means you get to see the opinions and views of others to get a better sense of the space here and how people think about it.

In this case, i'm sharing my perspective (both as a team and LDM member) of why i don't think this is an important space to focus on.
